### PR TITLE
feat: Add support to do kubernetes node monitoring

### PR DESCRIPTION
**Pull request by slashpai**
Related-to https://github.com/prometheus-operator/prometheus-operator/issues/5711

In this PR support is only added for kubernetes node monitoring

This is still in draft but I could do minimal monitoring with this.

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: ScrapeConfig
metadata:
  name: scrape-config-kubernetes-sd-example
  namespace: default
  labels:
    app.kubernetes.io/name: scrape-config-kubernetes-sd-example
spec:
  scheme: HTTPS
  authorization:
    credentials:
      name: prometheus-secret
      key: token
  tlsConfig:
    ca:
      secret:
        name: default-server
        key: ca.crt
    insecureSkipVerify: true
  kubernetesSDConfigs:
  - role: Node
```

## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Add Kubernetes node monitoring support through ScrapeConfig CRD
```

### CODE REVIEW

File: pkg/prometheus/promcfg.go
Comment (slashpai) [inline pkg/prometheus/promcfg.go:44]: My idea was to use this but I am letting specify role in ScrapeConfig spec. I will revisit this

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:172]: (nit) to comply with the Kubernetes API [conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#constants), constant values should be CamelCase.
```suggestion
	//+kubebuilder:validation:Enum=Node
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:103]: see below
```suggestion
	// +kubebuilder:validation:Enum=HTTP;HTTPS
```

**Review by simonpasquier — COMMENTED** [review]
I'd start with adding HTTPS support for scrape configs to ease the review (and it's also a useful feature on its own).

**Comment by slashpai** [comment]
> I'd start with adding HTTPS support for scrape configs to ease the review (and it's also a useful feature on its own).

Sure I will move, I had thought same I needed the change to test kubernetes node

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go
Comment (slashpai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:103]: setting with capitals, prometheus throws error

```
ts=2023-06-15T06:03:34.480Z caller=scrape.go:499 level=error component="scrape manager" scrape_pool=scrapeconfig/default/scrape-config-scheme msg="Creating target failed" err="instance 0 in group 0: invalid scheme: \"HTTPS\""
```

Comment (slashpai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:103]: Should we convert this to lowercase in operator code? I found api validation not fully useful if we do have to convert again in code

Comment (slashpai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:103]: addressed here https://github.com/prometheus-operator/prometheus-operator/pull/5679#discussion_r1230552798

**Comment by slashpai** [comment]
@simonpasquier 

I tested with following config and it works, without tlsconfig, auth and scheme it gives Bad Request error even if we run operator inside cluster. Let me know if this is ok so I can add test for same

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: ScrapeConfig
metadata:
  name: scrape-config-kubernetes-sd-example
  namespace: default
  labels:
    app.kubernetes.io/name: scrape-config-kubernetes-sd-example
spec:
  scheme: HTTPS
  authorization:
    credentials:
      name: prometheus-secret
      key: token
  tlsConfig:
    ca:
      secret:
        name: default-server
        key: ca.crt
    insecureSkipVerify: true
  kubernetesSDConfigs:
  - role: Node
```

**Comment by slashpai** [comment]
I am trying to write one more e2e test

File: test/framework/framework.go
Comment (slashpai) [inline test/framework/framework.go:592]: cluster role binding was required for `prometheus` so that `prometheus serviceAccount` has access to list Node objects otherwise prometheus errors following

```
ts=2023-07-03T05:47:22.289Z caller=klog.go:108 level=warn component=k8s_client_runtime func=Warningf msg="pkg/mod/k8s.io/client-go@v0.26.2/tools/cache/reflector.go:169: failed to list *v1.Node: nodes is forbidden: User \"system:serviceaccount:allns-y-scrapeconfigkubenode-rx7h5h-0-9252af28:prometheus\" cannot list resource \"nodes\" in API group \"\" at the cluster scope"
ts=2023-07-03T05:47:22.289Z caller=klog.go:116 level=error component=k8s_client_runtime func=ErrorDepth msg="pkg/mod/k8s.io/client-go@v0.26.2/tools/cache/reflector.go:169: Failed to watch *v1.Node: failed to list *v1.Node: nodes is forbidden: User \"system:serviceaccount:allns-y-scrapeconfigkubenode-rx7h5h-0-9252af28:prometheus\" cannot list resource \"nodes\" in API group \"\" at the cluster scope"
```

**Comment by slashpai** [comment]
> I am trying to write one more e2e test

This is done

**Review by JoaoBraveCoding — COMMENTED** [review]
some minor nits but the content lgtm

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go
Comment (nicolastakashi) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:83]: nit: you have two spaces here.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go
Comment (nicolastakashi) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:194]: Do we plan to support other kinds of roles like pod|service?

Comment (slashpai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:194]: I believe that would be future so that we don't replicate fields in all monitors/probe for new prometheus added features. 
May be once ScrapeConfig is v1 we can add new features only to ScrapeConfig CRD or we need to come up with migration plan to move monitors to ScrapeConfig. But it needs further discussions.

But thats my understanding from proposal https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/proposals/202212-scrape-config.md :)

File: test/e2e/scrapeconfig_test.go
Comment (nicolastakashi) [reply test/e2e/scrapeconfig_test.go:267]: I'd suggest something like that:
```
For Prometheus scrape nodes, it must use mTLS authentication, using the certificates issued by the Prometheus Service account```

**Comment by slashpai** [comment]
Addressed the comments

File: test/framework/framework.go
Comment (ArthurSens) [inline test/framework/framework.go:592]: Do we need to add a deletion step in the function `DeletePrometheusOperatorClusterResource` to delete the ClusterRoleBinding?

I noticed we don't have test failures in the PR so I'm not sure, but I'd expect that we would need to delete it 🤔

File: test/e2e/scrapeconfig_test.go
Comment (ArthurSens) [inline test/e2e/scrapeconfig_test.go:318]:
```suggestion
	// Remove the ScrapeConfig
```

You're checking that the targets disappeared in the next step, right? Let's not duplicate the comments :P

**Review by ArthurSens — APPROVED** [review]
Looks good to me :)

I think we'll need to make sure to document the RBAC requirements in https://github.com/prometheus-operator/prometheus-operator/pull/5659 as well

File: test/e2e/scrapeconfig_test.go
Comment (slashpai) [reply test/e2e/scrapeconfig_test.go:318]: aah good catch :D

File: test/framework/framework.go
Comment (slashpai) [reply test/framework/framework.go:592]: Ya I see that we could delete it. I just followed similar to RoleBinding :)

**Comment by slashpai** [comment]
Addressed Arthur comments

File: test/framework/framework.go
Comment (simonpasquier) [reply test/framework/framework.go:592]: I see that `SetupPrometheusRBACGlobal()` below configures the same rbac permissions. Can't we use it? I'm worried that changing the meaning of `SetupPrometheusRBAC()` hides problems in the long run (I wouldn't mind additional comments on the `Setup*RBAC*` functions to clarify what they do).

File: test/e2e/prometheus_test.go
Comment (xiu) [inline test/e2e/prometheus_test.go:60]: (nit) should it be `createMTLSSecret?

**Review by xiu — APPROVED** [review]
Added a comment but LGTM otherwise, thanks! :)

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go
Comment (slashpai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: This is addressed

File: test/e2e/scrapeconfig_test.go
Comment (slashpai) [reply test/e2e/scrapeconfig_test.go:280]: On kind cluster I tried, 

```Get "https://172.18.0.8:10250/metrics": tls: failed to verify certificate: x509: cannot validate certificate for 172.18.0.8 because it doesn't contain any IP SANs```

File: test/e2e/prometheus_test.go
Comment (slashpai) [reply test/e2e/prometheus_test.go:60]: Since its usually abbreviated `mTLS` I used same but can change to this :)

File: test/e2e/scrapeconfig_test.go
Comment (simonpasquier) [reply test/e2e/scrapeconfig_test.go:283]: We discussed offline with @slashpai but in this case, it's ok to discard TLS verification.

**Comment by slashpai** [comment]
addressed comments

**Comment by slashpai** [comment]
I have squashed commits and addressed comments

@xiu / @ArthurSens can you have another review on this for merging :) ?

**Comment by xiu** [comment]
LGTM! If @ArthurSens is good with it, we can merge 👍

**Review by ArthurSens — APPROVED** [review]
Awesome work! 

Really good to see the whole effort broken down into smaller pieces and you following up until completion 🙂

**Pull request merged**
